lib/Conversion/ImportVerilog/ImportVerilog.cpp

```diff
@@ -390,3 +392,22 @@ void circt::registerFromVerilogTranslation() {
         return module;
       });
 }
+
+void circt::registerCanonicalizePasses() {
+  static TranslateToMLIRRegistration optimize(
+      "canonicalize", "import Verilog or SystemVerilog and optimize IR",
+      [](llvm::SourceMgr &sourceMgr, MLIRContext *context) {
+        TimingScope ts;
+        OwningOpRef<ModuleOp> module(
+            ModuleOp::create(UnknownLoc::get(context)));
+        if (failed(importVerilog(sourceMgr, context, ts, module.get())))
+          module = {};
+        mlir::PassManager pm(context);
+        pm.addPass(createCanonicalizerPass());
+        pm.addPass(createCSEPass());
+        if (failed(pm.run(*module)))
+          module = {};
+
+        return module;
+      });
+}
```

Comment:
I would not register the canonicalizer as a translation pass. Translation passes in MLIR are intended to convert from something outside MLIR into IR ops, and from IR ops to something outside MLIR. Examples here are other assembly text formats like FIRRTL, C/C++ code for the `emitc` dialect, LLVM IR, SPIR-V IR, etc. The canonicalizer however is an MLIR-to-MLIR transformation pass, which is already registered with `circt-opt`. You can just call `circt-opt --canonicalize` and get a canonicalized output :slightly_smiling_face:

I like your idea of starting to build up an entire Verilog pipeline though. You could move the `addPass()` calls to the `circt-verilog` tool (take a look at the `tools/circt-verilog/circt-verilog.cpp` file. That could be a great place for them!

Reply:
Thank you for the suggestion. I've separated the pass and translate, and moved `addPass` to the `circt-verilog` tool.